parser.lisp

```diff
@@ -20,26 +20,111 @@
 (define-matcher grouper (is #\,))
 (define-matcher attr-comparator (or (is #\=) (and (any #\~ #\^ #\$ #\* #\|) (next (is #\=)))))
 (defvar *valid-combinators* " >+~")
 
+(defun escapable (char)
+  "A helper function to decide whether a part of identifier needs escaping."
+  (let ((code (char-code char)))
+    (or (<= code #x1f)
+        (= code #x7f)
+        (not (or (>= code #x80)
+                 (char= char #\-)
+                 (char= char #\_)
+                 (digit-char-p char)
+                 (char<= #\A char #\Z)
+                 (char<= #\a char #\z))))))
+
+;; https://drafts.csswg.org/cssom/#serialize-an-identifier
+(defun css-escape (string)
+  "Escape all the invalid CSS characters to their safe counterparts."
+  (declare (optimize speed)
+           (type string string))
+  (if (or (some #'escapable string)
+          (and (= 1 (length string))
+               (char= #\- (elt string 0)))
+          (and (>= (length string) 1)
+               (digit-char-p (elt string 0)))
+          (and (>= (length string) 2)
+               (char= #\- (elt string 0))
+               (digit-char-p (elt string 1))))
+      (with-output-to-string (s)
+        (flet ((escape-regular-string (string)
+                 (dotimes (index (length string))
+                   (let ((char (elt string index)))
+                     (cond
+                       ((char= #\Nul char)
+                        (write-string "\\fffd " s))
+                       ((or (<= #x1 (char-code char) #x1f)
+                            (= (char-code char) #x7f))
+                        (format s "\\~X " (char-code char)))
+                       ((escapable char)
+                        (write-char #\\ s)
+                        (write-char char s))
+                       (t
+                        (write-char char s)))))))
+          ;; Process first char.
+          (cond
+            ((and (= (length string) 1)
+                  (char= #\- (elt string 0)))
+             (write-char #\\ s)
+             (write-char (elt string 0) s))
+            ((and (= (length string) 1)
+                  (digit-char-p (elt string 0)))
+             (format s "\\~X " (char-code (elt string 0))))
+            (t
+             (escape-regular-string (subseq string 0 1))))
+          ;; Second char.
+          (when (>= (length string) 2)
+            (if (and (char= #\- (elt string 0))
+                     (digit-char-p (elt string 1)))
+                (format s "\\~X " (char-code (elt string 1)))
+                (escape-regular-string (subseq string 1 2))))
+          ;; All the rest.
+          (when (> (length string) 2)
+            (escape-regular-string (subseq string 2)))))
+      string))
+
+(defun css-unescape (string)
+  "Get the original contents of the escaped STRING."
+  (declare (optimize speed)
+           (type string string))
+  (if (search "\\" string)
+      (with-output-to-string (s)
+        (do* ((index 0 (1+ index))
```

Comment:
To be clear with my previous comment I did not mean to say that you shouldn't use `loop`. Just that you should use `(loop ... do (cond ..))` rather than `(loop ... if ..)`

Reply:
So, sorry to make you run in circles, but please return to using `loop`.

Reply:
But I like `(loop ... if ...)` :(

Anyway, I've rewritten it as `(loop ... do (cond ...))` now, and it works just fine.